### ISSUE TITLE
(staging) systemd fails to cross compile to arm

### ISSUE THREAD

**Issue by sorki** [NixOS/nixpkgs#86644]
**Describe the bug**

Building `systemd-boot` with `EFI` support uses host `cc` and `ld` correctly but with target arch set to `arm` resulting in

```
/nix/store/k930l9lckxjf4zmafv4aapvdbd2sibmj-gcc-wrapper-9.3.0/bin/ld -o src/boot/efi/systemd_boot.so -T /nix/store/zkghrargax07rq16d6wa4r68wa1d7p30-gnu-efi-3.0.11-armv7l-unknown-linux-gnueabihf/lib/elf_arm_efi.lds -shared -Bsymbolic -nostdlib -znocombreloc -L /nix/store/zkghrargax07rq16d6wa4r68wa1d7p30-gnu-efi-3.0.11-armv7l-unknown-linux-gnueabihf/lib /nix/store/zkghrargax07rq16d6wa4r68wa1d7p30-gnu-efi-3.0.11-armv7l-unknown-linux-gnueabihf/lib/crt0-efi-arm.o --defsym=EFI_SUBSYSTEM=0xa src/boot/efi/disk.c.o src/boot/efi/graphics.c.o src/boot/efi/measure.c.o src/boot/efi/pe.c.o src/boot/efi/util.c.o src/boot/efi/boot.c.o src/boot/efi/console.c.o src/boot/efi/crc32.c.o src/boot/efi/random-seed.c.o src/boot/efi/sha256.c.o src/boot/efi/shim.c.o -lefi -lgnuefi /nix/store/ppbczakh03k4ag97v966bh5aw6ps7dqr-armv7l-unknown-linux-gnueabihf-stage-final-gcc-debug-9.3.0/lib/gcc/armv7l-unknown-linux-gnueabihf/9.3.0/libgcc.a
/nix/store/3b3ighb83nhifa1v4n7855hlbdl1mhf9-binutils-2.31.1/bin/ld: cannot represent machine `arm'
```

Workaround is to disable `EFI` completely by adding the following to `mesonFlags`
```
(if stdenv.hostPlatform != stdenv.targetPlatform then "-Defi=false" else "")
```

but I would like to fix this correctly.

There's also `-Defi-ld` flag which can be used to pass another `ld` but I'm not sure if using target's `ld` is the right approach here.

**To Reproduce**

Cross-build to `armv7l` with current staging.

```
nixpkgs.crossSystem.system = "armv7l-linux";
```

**Notify maintainers**

Ping cross gurus @Mic92 @Ericson2314 @matthewbauer

**Comment by sorki** [comment]
Hm, the workaround doesn't seem to be enough and I've switched it to

```
    #(if !stdenv.hostPlatform.isEfi then "-Dgnu-efi=false" else "-Dgnu-efi=true")
    "-Defi=false"
```

**Comment by sorki** [comment]
Looks like #72802 adds support for `withEfi` flag.

**Comment by russetrob** [comment]
Firstly, I assume you did check the resulting package runs on arm? I'd be suspicious it's not cross-compiling at all!
You could approach the problem the other way, and get efi linked into the arm build.
For this you'd need to find out why your linker can't understand "arm" as machine name. For instance on a mingw build of ld to cross compile to arm, ld accepts machine of "arm".
You might need to hack the linker command to take option "-marmelf".
It might help take the linker and run "/nix/store/k930l9lckxjf4zmafv4aapvdbd2sibmj-gcc-wrapper-9.3.0/bin/ld --help"  the output should specify emulations and output formats supported (Probably a good idea to paste it here, for cross-compile experts to mull over).

**Comment by sorki** [comment]
Thanks, the only problem is the EFI (systemd-boot) part, it does cross-compile and works correctly it without it. 

Trouble (for me :)) is expressing your suggestion using `pkgs.buildPackages` or similar. Will take a look at hacking the linker option but my suspicion is that it's not using the correct one

**Comment by sorki** [comment]
Fixed by a239864fc610ff7738c5a8af2a7be25a0af4a28b
